zebra-rpc/src/methods/get_block_template_rpcs.rs

```diff
@@ -904,6 +910,58 @@ where
         }
         .boxed()
     }
+
+    fn get_difficulty(&self) -> BoxFuture<Result<f64>> {
+        let network = self.network;
+        let mut state = self.state.clone();
+
+        async move {
+            let request = ReadRequest::ChainInfo;
+
+            let response = state
+                .ready()
+                .and_then(|service| service.call(request))
+                .await
+                .map_err(|error| Error {
+                    code: ErrorCode::ServerError(0),
+                    message: error.to_string(),
+                    data: None,
+                })?;
+
+            let chain_info = match response {
+                ReadResponse::ChainInfo(info) => info,
+                _ => unreachable!("unmatched response to a chain info request"),
+            };
+
+            // The following code is ported from zcashd implementation.
+            // https://github.com/zcash/zcash/blob/v5.4.0-rc4/src/rpc/blockchain.cpp#L46-L73
+
+            let pow_limit = u32::from_be_bytes(
+                zebra_chain::work::difficulty::ExpandedDifficulty::target_difficulty_limit(network)
+                    .to_compact()
+                    .bytes_in_display_order(),
+            );
+            let bits = u32::from_be_bytes(chain_info.expected_difficulty.bytes_in_display_order());
+
+            let mut n_shift = (bits >> 24) & 0xff;
+            let n_shift_amount = (pow_limit >> 24) & 0xff;
+
+            let mut d_diff: f64 = (pow_limit & 0x00ffffff) as f64 / (bits & 0x00ffffff) as f64;
+
+            while n_shift < n_shift_amount {
+                d_diff *= 256.0;
+                n_shift += 1;
+            }
+
+            while n_shift > n_shift_amount {
+                d_diff /= 256.0;
+                n_shift -= 1;
+            }
```

Comment:
This looks like a re-implementation of `CompactDifficulty::to_expanded()`, but on `f64`, rather than `U256`:
https://github.com/ZcashFoundation/zebra/blob/d67b3b641d2a9bb5680c91cd3281bddf40670c42/zebra-chain/src/work/difficulty.rs#L148-L156

I think we can do this in a simpler way using our existing code, see PR #6105 for my suggestion.

This helps us avoid duplicate implementations of the consensus rules.

Reply:
I agree that your code is simpler but i don't see the consensus rules here, this is just for displaying (no `# Consensus` doc or similar was added here).

The problem i see with the new code is the diff which seems to be bigger than the one from this code. 

The extra tests are fine.

Reply:
> I agree that your code is simpler but i don't see the consensus rules here, this is just for displaying (no `# Consensus` doc or similar was added here).

The code isn't consensus-critical, but it uses the same consensus rules to display an accurate difficulty value to users.

So maybe there could be a doc in this PR that quotes the `ToTarget` consensus rules:
https://zips.z.cash/protocol/protocol.pdf#nbits

Or for PR #6105, we should turn these existing comments into consensus docs, because that's the consensus implementation we're using for the RPC and in other code that is actually consensus-critical:
https://github.com/ZcashFoundation/zebra/blob/7c36ff611fc9ad7fd5d826aa8380a0290ed5faab/zebra-chain/src/work/difficulty.rs#L96
https://github.com/ZcashFoundation/zebra/blob/7c36ff611fc9ad7fd5d826aa8380a0290ed5faab/zebra-chain/src/work/difficulty.rs#L318

Reply:
Just to be clear, this isn't a blocker, feel free to merge this PR and we can fix it up in the next PR.